### PR TITLE
Disable build of HWAES on PPC Macs

### LINKED ISSUE #22818: OpenSSL 3.2.0 build fails on 32bit PowerPC macs running OS X
On a 32bit (ppc7450) PowerPC Mac (iBook G4) running Mac OS X 10.5 (darwin 9), the build of OpenSSL 3.2.0 fails due to the use of 64bit instructions in `crypto/aes/aesp8-ppc.s` which I guess is for POWER8 but targets ppc970 (G5) which is 64bit.

I guess we now need to state `no-asm` on such systems?

Invoking the build with 
```
perl Configure --prefix=/tmp/openssl3 no-ssl3 no-ssl3-method no-zlib darwin-ppc-cc
```

Build eventually fails with
```
cc -fPIC -fno-common -arch ppc -std=gnu9x -Wa,-force_cpusubtype_ALL -O3 -DB_ENDIAN -DOPENSSL_PIC -DOPENSSLDIR="\"/tmp/openssl3/ssl\"" -DENGINESDIR="\"/tmp/openssl3/lib/engines-3\"" -DMODULESDIR="\"/tmp/openssl3/lib/ossl-modules\"" -D_REENTRANT -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -c -o crypto/aes/libcrypto-shlib-aesp8-ppc.o crypto/aes/aesp8-ppc.s
crypto/aes/aesp8-ppc.s:2343:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2345:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2350:Invalid mnemonic 'lxvw4x'
crypto/aes/aesp8-ppc.s:2411:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2412:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:2421:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2422:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:2432:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2433:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:2443:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2444:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:2454:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2455:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:2465:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2466:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:2492:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2510:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2511:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:2527:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2528:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:2549:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2550:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:2566:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2567:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:2583:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2584:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:2606:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2607:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2608:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:2609:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:2642:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3021:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3023:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3028:Invalid mnemonic 'lxvw4x'
crypto/aes/aesp8-ppc.s:3081:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3082:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:3091:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3092:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:3102:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3103:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:3113:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3114:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:3124:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3125:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:3135:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3136:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:3162:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3180:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3181:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:3197:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3198:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:3219:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3220:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:3236:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3237:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:3253:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3254:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:3276:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3277:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3278:Invalid mnemonic 'vpermxor'
crypto/aes/aesp8-ppc.s:3279:Invalid mnemonic 'xxlor'
crypto/aes/aesp8-ppc.s:3310:Invalid mnemonic 'xxlor'
```

Config dump

```
Command line (with current working directory = .):

    /Users/sme/tigerbrew/opt/perl/bin/perl ./Configure --prefix=/tmp/openssl3 no-ssl3 no-ssl3-method no-zlib darwin-ppc-cc

Perl information:

    /Users/sme/tigerbrew/opt/perl/bin/perl
    5.38.0 for darwin-thread-multi-2level

Enabled features:

    apps
    argon2
    aria
    asm
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    bulk
    cached-fetch
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    default-thread-pool
    deprecated
    des
    dgram
    dh
    docs
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    ecx
    engine
    err
    filenames
    gost
    http
    idea
    legacy
    loadereng
    makedepend
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    ocb
    ocsp
    padlockeng
    pic
    pinshared
    poly1305
    posix-io
    psk
    quic
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    shared
    siphash
    siv
    sm2
    sm2-precomp
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    ssl-trace
    static-engine
    stdio
    tests
    thread-pool
    threads
    tls
    ts
    ui-console
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    acvp-tests          [cascade]             OPENSSL_NO_ACVP_TESTS
    afalgeng            [not-linux]           OPENSSL_NO_AFALGENG
    asan                [default]             OPENSSL_NO_ASAN
    brotli              [default]             OPENSSL_NO_BROTLI
    brotli-dynamic      [default]             OPENSSL_NO_BROTLI_DYNAMIC
    buildtest-c++       [default]             
    winstore            [not-windows]         OPENSSL_NO_WINSTORE
    crypto-mdebug       [default]             OPENSSL_NO_CRYPTO_MDEBUG
    devcryptoeng        [default]             OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128 [default]             OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                 [default]             OPENSSL_NO_EGD
    external-tests      [default]             OPENSSL_NO_EXTERNAL_TESTS
    fips                [default]             
    fips-securitychecks [cascade]             OPENSSL_NO_FIPS_SECURITYCHECKS
    fuzz-afl            [default]             OPENSSL_NO_FUZZ_AFL
    fuzz-libfuzzer      [default]             OPENSSL_NO_FUZZ_LIBFUZZER
    ktls                [default]             OPENSSL_NO_KTLS
    md2                 [default]             OPENSSL_NO_MD2 (skip crypto/md2)
    msan                [default]             OPENSSL_NO_MSAN
    rc5                 [default]             OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                [default]             OPENSSL_NO_SCTP
    tfo                 [default]             OPENSSL_NO_TFO
    trace               [default]             OPENSSL_NO_TRACE
    ubsan               [default]             OPENSSL_NO_UBSAN
    unit-test           [default]             OPENSSL_NO_UNIT_TEST
    uplink              [no uplink_arch]      OPENSSL_NO_UPLINK
    weak-ssl-ciphers    [default]             OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                [option]              OPENSSL_NO_ZLIB
    zlib-dynamic        [default]             OPENSSL_NO_ZLIB_DYNAMIC
    zstd                [default]             OPENSSL_NO_ZSTD
    zstd-dynamic        [default]             OPENSSL_NO_ZSTD_DYNAMIC
    ssl3                [option(ssl3-method)] OPENSSL_NO_SSL3
    ssl3-method         [option]              OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "ar",
    ARFLAGS => "qc",
    CC => "cc",
    CFLAGS => "-O3",
    HASHBANGPERL => "/usr/bin/env perl",
    RANLIB => "ranlib -c",
    RC => "windres",
    asm_arch => "ppc32",
    bn_ops => "BN_LLONG RC4_CHAR",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cflags => "-arch ppc -std=gnu9x -Wa,-force_cpusubtype_ALL",
    cppflags => "-D_REENTRANT",
    defines => [ "OPENSSL_BUILDING_OPENSSL" ],
    disable => [  ],
    dso_scheme => "dlfcn",
    enable => [  ],
    ex_libs => "",
    includes => [  ],
    lflags => "-Wl,-search_paths_first",
    lib_cflags => "",
    lib_cppflags => "-DB_ENDIAN",
    lib_defines => [  ],
    module_cflags => "-fPIC -fno-common",
    module_cxxflags => undef,
    module_ldflags => "-bundle",
    perl_platform => "Unix",
    perlasm_scheme => "osx32",
    shared_cflag => "-fPIC -fno-common",
    shared_defines => [  ],
    shared_extension => ".\$(SHLIB_VERSION_NUMBER).dylib",
    shared_ldflag => "-dynamiclib -current_version \$(SHLIB_VERSION_NUMBER) -compatibility_version \$(SHLIB_VERSION_NUMBER)",
    shared_rcflag => "",
    shared_sonameflag => "-install_name \$(libdir)/",
    shared_target => "darwin-shared",
    sys_id => "MACOSX",
    thread_defines => [  ],
    thread_scheme => "pthreads",
    unistd => "<unistd.h>",

Recorded environment:

    AR = 
    ARFLAGS = 
    AS = 
    ASFLAGS = 
    BUILDFILE = 
    CC = 
    CFLAGS = 
    CPP = 
    CPPDEFINES = 
    CPPFLAGS = 
    CPPINCLUDES = 
    CROSS_COMPILE = 
    CXX = 
    CXXFLAGS = 
    HASHBANGPERL = 
    LD = 
    LDFLAGS = 
    LDLIBS = 
    MT = 
    MTFLAGS = 
    OPENSSL_LOCAL_CONFIG_DIR = 
    PERL = 
    RANLIB = 
    RC = 
    RCFLAGS = 
    RM = 
    WINDRES = 
    __CNF_CFLAGS = 
    __CNF_CPPDEFINES = 
    __CNF_CPPFLAGS = 
    __CNF_CPPINCLUDES = 
    __CNF_CXXFLAGS = 
    __CNF_LDFLAGS = 
    __CNF_LDLIBS = 

Makevars:

    AR              = ar
    ARFLAGS         = qc
    ASFLAGS         = 
    CC              = cc
    CFLAGS          = -O3
    CPPDEFINES      = 
    CPPFLAGS        = 
    CPPINCLUDES     = 
    CXXFLAGS        = 
    HASHBANGPERL    = /usr/bin/env perl
    LDFLAGS         = 
    LDLIBS          = 
    PERL            = /Users/sme/tigerbrew/opt/perl/bin/perl
    RANLIB          = ranlib -c
    RC              = windres
    RCFLAGS         = 

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    Makefile

build file:

    Makefile

build file templates:

    Configurations/common0.tmpl
    Configurations/unix-Makefile.tmpl
```

### PR DESCRIPTION
**Pull request by t8m**
Fixes #22818

### CODE REVIEW

**Comment by kroeckx** [comment]
Shouldn't it just move to the AESASM_ppc64?

**Comment by t8m** [comment]
> Shouldn't it just move to the AESASM_ppc64?

No because
1. aesp8-ppc.pl can be used on 32 bit ppc builds if the assembler understands the instructions and the CPU is detected to support Power ISA 2.07 
2. aesp8-ppc.pl now cannot be used on 64 bit ppc builds if the assembler does not understand the vpermxor instructions

**Comment by kroeckx** [comment]
Does this disable assembler that was previously used on those platforms?

**Comment by t8m** [comment]
> Does this disable assembler that was previously used on those platforms?

I do not think so - at least for Macs. The assembler was compiled in but it was never invoked anyway AFAIK.

AIX might be a different story - I am just replicating what we had to do with ChaCha20.

File: include/crypto/aes_platform.h [hunk 1]
```diff
@@ -65,16 +65,16 @@ void AES_xts_decrypt(const unsigned char *inp, unsigned char *out, size_t len,
...
-#   define HWAES_CAPABLE  (OPENSSL_ppccap_P & PPC_CRYPTO207)
-#   define HWAES_set_encrypt_key aes_p8_set_encrypt_key
-#   define HWAES_set_decrypt_key aes_p8_set_decrypt_key
-#   define HWAES_encrypt aes_p8_encrypt
-#   define HWAES_decrypt aes_p8_decrypt
-#   define HWAES_cbc_encrypt aes_p8_cbc_encrypt
-#   define HWAES_ctr32_encrypt_blocks aes_p8_ctr32_encrypt_blocks
-#   define HWAES_xts_encrypt aes_p8_xts_encrypt
-#   define HWAES_xts_decrypt aes_p8_xts_decrypt
 #   if !defined(OPENSSL_SYS_AIX) && !defined(OPENSSL_SYS_MACOSX)
```
Comment (nhorman) [inline include/crypto/aes_platform.h:68]: IIRC some Darwin systems were 64 bit PPC.  Is this going to disable the asm support for those platforms?  Or is that acceptable here?

Comment (t8m) [reply include/crypto/aes_platform.h:68]: No, it won't disable the asm support for AES completely. It's just the PowerISA 2.07 code that won't be built. And AFAIK it was not used anyway.

Comment (tmshort) [reply include/crypto/aes_platform.h:68]: PowerISA 2.07 code won't be built for 32-bit and 64-bit; so _that_ code is gone from macOS regardless of whether its supported (i.e. on 32-bit)?

**Comment by tmshort** [comment]
This brings up something interesting... will support for PPC OS X be dropped at any point in the near future? Apple has already dropped support as of OS X 10.6 (Snow Leopard). OS X 10.5.8 was the last version to support PPC, and that was released 2009-08-13.

File: include/crypto/aes_platform.h (hunk 1 above)
Comment (t8m) [reply include/crypto/aes_platform.h:68]: @tmshort AFAIK it was never supported by the CPUs used on Macs? Its just that the assembler previously built but it was not used anyway.

**Review by nhorman — APPROVED** [review]
LGTM

**Review by tmshort — APPROVED** [review]
LGTM
I can't test it out, as I no longer have a PPC Mac.

**Comment by kroeckx** [comment]
So my understanding of the issue is:
- Some assemblers don't understand the instructions. We can either convert it to opcodes ourself, or just not build it for some platforms.
- The code would only be used on CPUs that support it, including a 32 bit build for 64 bit CPUs.
- At least the Mac never supported on a CPU that can make use of this

So I think for AIX, we have 2 options:
- Don't build it
- Convert it ourself

I think we've done both things in the past. It might be useful to have some kind of policy for this.

I think that if we decide to go with disabling things on a platform that can support it, we should at least document so.

**Comment by t8m** [comment]
This is a quick fix for the regression. We would need a proper community maintainer for AIX platform to do any better fix there. I suggest we merge this as is. Or alternatively I drop AIX from this patch as we did not get any bug report from AIX users yet, although I highly suspect we will get them later.

**Comment by t8m** [comment]
ping for an OTC review

File: include/crypto/aes_platform.h (hunk 1 above)
Comment (barracuda156) [reply include/crypto/aes_platform.h:68]: G5 is ISA 2.02 with some extensions, AFAIK. Anything above ISA 2.03 is off question for sure.

**Review by mattcaswell — APPROVED** [review]

**Comment by barracuda156** [comment]
> This brings up something interesting... will support for PPC OS X be dropped at any point in the near future? Apple has already dropped PPC support as of OS X 10.6 (Snow Leopard). OS X 10.5.8 was the last version to support PPC, and that was released 2009-08-13. (Snow Leopard 10.6 did support 32-bit PPC apps via Rosetta, 10.6.8 was released 2011-07-25.)

For the record, this is not strictly accurate. While 10.6.8 only supports `ppc` via Rosetta, there are early versions of 10.6 running natively on PowerPC – in fact, that is my main testing system.

**Comment by barracuda156** [comment]
@t8m I am sorry, perhaps I miss something, but the header seems not to have `ppc64` case at all, how is it supposed to work? Without `__ppc64__` it gonna be omitted on Darwin when building for ppc64.

**Comment by t8m** [comment]
> @t8m I am sorry, perhaps I miss something, but the header seems not to have `ppc64` case at all, how is it supposed to work? Without `__ppc64__` it gonna be omitted on Darwin when building for ppc64.
> 
> I refer to this:
> 
> https://github.com/openssl/openssl/blob/b062a3c552bf283319dede3437598f1747730053/include/crypto/aes_platform.h#L63

Hmm... that is interesting. Not sure if this ever worked properly. However this is a separate issue. If you want to, please report a new issue. We would need someone to verify that it actually builds and works correctly on 64 bit PPC builds and it would be a master branch only fix.

**Comment by barracuda156** [comment]
@t8m I have opened an issue for `ppc64`: https://github.com/openssl/openssl/issues/23220

> I do not think so - at least for Macs. The assembler was compiled in but it was never invoked anyway AFAIK.

Could you advise how to test that locally? It would be really undesirable to remove support for something which worked earlier, provided it did in fact work. If it never worked to begin with, then of course disabling it is the right way to go, until someone bother to re-implement it it a functional way.

**Comment by barracuda156** [comment]
> > Could you advise how to test that locally?
> 
> You would have to set a breakpoint to that code (or patch it so it aborts) to see if it was invoked.

And what will be a test case where it should get invoked if working? Or just running make test?

**Comment by t8m** [comment]
Running `util/shlib_wrap.sh test/evp_test test/recipes/30-test_evp_data/evpciph_aes_common.txt` should invoke enough AES to see whether it calls this assembler code or not.

**Comment by t8m** [comment]
Merged to the master and 3.2 branches. Thank you for the reviews.

**Comment by barracuda156** [comment]
@t8m Sorry, I did not give up on checking assembler, I just had no time to deal with this yesterday. I will return to the matter shortly.

**Pull request closed without merging**